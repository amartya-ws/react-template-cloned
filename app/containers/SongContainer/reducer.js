/*
 *
 * HomeContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const { Types: songContainerTypes, Creators: songContainerCreators } = createActions({
  requestGetItuneSongs: ['artistName'],
  successGetItuneSongs: ['data'],
  clearItuneSongs: {}
});
export const initialState = { artistName: null, ituneData: {} };

/* eslint-disable default-case, no-param-reassign */
export const songContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case songContainerTypes.REQUEST_GET_ITUNE_SONGS:
        draft.artistName = action.artistName;
        break;
      case songContainerTypes.SUCCESS_GET_ITUNE_SONGS:
        draft.ituneData = action.data;
        break;
      case songContainerTypes.CLEAR_ITUNE_SONGS:
        draft.artistName = null;
        draft.ituneData = {};
        break;
    }
  });

export default songContainerReducer;
