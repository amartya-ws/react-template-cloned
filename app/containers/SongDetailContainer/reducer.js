/*
 *
 * SongDetail reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = {
  trackId: null,
  singleItune: {}
};

export const { Types: songDetailContainerTypes, Creators: songDetailContainerCreators } = createActions({
  requestGetSingleItune: ['trackId'],
  successGetSingleItune: ['data']
});

/* eslint-disable default-case, no-param-reassign */
export const songDetailContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case songDetailContainerTypes.REQUEST_GET_SINGLE_ITUNE:
        draft.trackId = action.trackId;
        break;
      case songDetailContainerTypes.SUCCESS_GET_SINGLE_ITUNE:
        draft.singleItune = action.data;
        break;
    }
  });

export default songDetailContainerReducer;
