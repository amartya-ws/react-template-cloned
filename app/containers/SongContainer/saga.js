import { put, call, takeLatest } from 'redux-saga/effects';
import { songContainerTypes, songContainerCreators } from './reducer';
import { getItunes } from '@app/services/ituneApi';

// Individual exports for testing
const { REQUEST_GET_ITUNE_SONGS } = songContainerTypes;
const { successGetItuneSongs } = songContainerCreators;

export function* getItuneSongs(action) {
  const response = yield call(getItunes, action.artistName);
  const { data, ok } = response;
  if (ok) {
    yield put(successGetItuneSongs(data));
  }
}

export default function* songContainerSaga() {
  yield takeLatest(REQUEST_GET_ITUNE_SONGS, getItuneSongs);
}
