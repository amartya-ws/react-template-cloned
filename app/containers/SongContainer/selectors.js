import { createSelector } from 'reselect';
import { initialState } from './reducer';
import get from 'lodash/get';

/**
 * Direct selector to the songContainer state domain
 */

export const selectSongContainerDomain = (state) => state.songContainer || initialState;

export const selectItuneData = () =>
  createSelector(selectSongContainerDomain, (substate) => get(substate, 'ituneData'));

export const selectArtistName = () =>
  createSelector(selectSongContainerDomain, (substate) => get(substate, 'artistName'));
