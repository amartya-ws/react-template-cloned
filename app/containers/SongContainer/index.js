/**
 *
 * SongContainer Container
 *
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { injectSaga } from 'redux-injectors';
import { Input } from 'antd';
import get from 'lodash/get';
import debounce from 'lodash/debounce';
import styled from 'styled-components';
import For from '@app/components/For';
import T from '@components/T';
import ItuneCard from '@app/components/ItuneCard';
import { selectItuneData, selectArtistName } from './selectors';
import { songContainerCreators } from './reducer';
import saga from './saga';

const StyledInput = styled(Input)`
  && {
    width: 25rem;
    height: 2.5rem;
    margin-bottom: 1rem;
  }
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const CardWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  gap: 1rem;
`;

export function SongContainer({ artistName, ituneData, dispatchItuneSongs, dispatchClearSongs }) {
  const [currentSongId, setCurrentSongId] = useState(0);
  const changeHandler = (searchTerm) => {
    if (searchTerm) {
      dispatchItuneSongs(searchTerm);
    } else {
      dispatchClearSongs();
    }
  };
  const debouncedHandler = debounce(changeHandler, 500);

  const renderSongList = () => {
    let songs = get(ituneData, 'results', null);
    return (
      <>
        <For
          of={songs}
          ParentComponent={CardWrapper}
          renderItem={(item, index) => (
            <ItuneCard key={index} {...item} currentSongId={currentSongId} setCurrentSongId={setCurrentSongId} />
          )}
        />
      </>
    );
  };

  return (
    <Wrapper>
      <T id="get_song_details" marginBottom={10} />
      <StyledInput
        placeholder="enter artist"
        defaultValue={artistName}
        type="text"
        data-testid="search-bar"
        onChange={(evt) => debouncedHandler(evt.target.value)}
      />
      <CardWrapper>{renderSongList()}</CardWrapper>
    </Wrapper>
  );
}

SongContainer.propTypes = {
  artistName: PropTypes.string,
  ituneData: PropTypes.object,
  dispatchClearSongs: PropTypes.func,
  dispatchItuneSongs: PropTypes.func
};

const mapStateToProps = createStructuredSelector({
  ituneData: selectItuneData(),
  artistName: selectArtistName()
});

export function mapDispatchToProps(dispatch) {
  const { requestGetItuneSongs, clearItuneSongs } = songContainerCreators;
  return {
    dispatchItuneSongs: (artistName) => dispatch(requestGetItuneSongs(artistName)),
    dispatchClearSongs: () => dispatch(clearItuneSongs())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, injectSaga({ key: 'songContainer', saga }))(SongContainer);

export const SongContainerTest = compose(injectIntl)(SongContainer);
