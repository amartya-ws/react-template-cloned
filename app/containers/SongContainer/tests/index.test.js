/**
 *
 * Tests for SongContainer container
 *
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderProvider, timeout } from '@utils/testUtils';
import { SongContainerTest as SongContainer } from '../index';
import { songContainerTypes } from '../reducer';
import { mapDispatchToProps } from '@app/containers/SongContainer/index';

describe('<SongContainer /> container tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<SongContainer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchItuneSongs on change', async () => {
    const artistName = 'eminem';
    const { getByTestId } = renderProvider(<SongContainer dispatchItuneSongs={submitSpy} />);
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: {
        value: artistName
      }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith(artistName);
  });

  it('should call dispatchClearSongs on empty change', async () => {
    let clearSongsSpy = jest.fn();
    let getItuneSongsSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <SongContainer dispatchClearSongs={clearSongsSpy} dispatchItuneSongs={getItuneSongsSpy} />
    );
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: {
        value: 'test'
      }
    });
    await timeout(500);
    expect(getItuneSongsSpy).toBeCalled();
    fireEvent.change(searchBar, {
      target: {
        value: ''
      }
    });
    await timeout(500);
    expect(clearSongsSpy).toBeCalled();
  });

  it('should validate mapDispatchToProps actions', async () => {
    const artistName = 'eminem';
    const actions = {
      dispatchItuneSongs: { type: songContainerTypes.REQUEST_GET_ITUNE_SONGS, artistName },
      dispatchClearSongs: { type: songContainerTypes.CLEAR_ITUNE_SONGS }
    };

    const props = mapDispatchToProps(submitSpy);
    props.dispatchItuneSongs(artistName);
    expect(submitSpy).toBeCalledWith(actions.dispatchItuneSongs);
    await timeout(500);

    props.dispatchClearSongs();
    expect(submitSpy).toBeCalledWith(actions.dispatchClearSongs);
  });

  it('should render exact number of ItuneCards as per totalCount in result', () => {
    const resultCount = 2;
    const ituneData = {
      resultCount,
      results: [
        {
          artistName: 'tupac',
          trackName: 'california love',
          artWorkUrl100:
            'https://is2-ssl.mzstatic.com/image/thumb/Music116/v4/35/8a/d6/358ad669-4087-07c2-e14e-4f4fd66ff588/094637927359.jpg/100x100bb.jpg'
        },
        {
          artistName: 'tupac',
          trackName: 'california love',
          artWorkUrl100:
            'https://is2-ssl.mzstatic.com/image/thumb/Music116/v4/35/8a/d6/358ad669-4087-07c2-e14e-4f4fd66ff588/094637927359.jpg/100x100bb.jpg'
        }
      ]
    };
    const { getAllByTestId } = renderProvider(<SongContainer ituneData={ituneData} dispatchItuneSongs={submitSpy} />);
    expect(getAllByTestId('itune-card').length).toBe(resultCount);
  });
});
