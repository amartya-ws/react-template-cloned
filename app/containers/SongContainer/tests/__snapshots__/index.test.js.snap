// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <h3>
        Get itune song details
      </h3>
      <input
        class="ant-input SongContainer__StyledInput-sc-1lfsiot-0 loJJPT"
        placeholder="enter artist"
        type="text"
        value=""
      />
    </div>
  </div>
</body>
`;
